GVFS/GVFS.Common/Http/HttpRequestor.cs

```diff
@@ -152,7 +153,15 @@ public void Dispose()
                     errorMessage = response.Content.ReadAsStringAsync().GetAwaiter().GetResult();
                     int statusInt = (int)response.StatusCode;
 
-                    if (response.StatusCode == HttpStatusCode.Unauthorized || response.StatusCode == HttpStatusCode.BadRequest || response.StatusCode == HttpStatusCode.Redirect)
+                    bool shouldRetry = ShouldRetry(response.StatusCode);
+
+                    if (response.StatusCode == HttpStatusCode.Unauthorized &&
+                        this.authentication.IsAnonymous)

```

Comment:
In that code above, if we're currently anonymous, then we don't call `TryGetCredentials` and so we remain anonymous. Then this code here says that we're anonymous and got back a 401, fail right away without retrying.

The general problem is that we must not call `git credential` if we want to use anonymous auth, because that will cause the GCM to show an auth prompt to the user. So this class here can't ever decide to call `git credential` on its own, since it is used in both anonymous and authenticated uses. It's really up to each entry point to attempt an anonymous call, and if that fails, force a call to both flip the IsAnonymous bit and also call `git credential` to generate/retrieve a PAT.

I do need to make a pass over this code and clean things up, because right now the usage is not totally clear. But we did confirm that these changes work well against a repo that allows anonymous access to its /gvfs/* endpoints.

Reply:
Arg! I was bitten by the auto-collapse of the file contents.  I thought that we were still inside the if block started on line 74.